Emby.Server.Implementations/LiveTv/Listings/XmlTvListingsProvider.cs

```diff
@@ -67,110 +64,65 @@ private async Task<string> GetXml(ListingsProviderInfo info, CancellationToken c
             }
+            else
+            {
+                Directory.CreateDirectory(Path.GetDirectoryName(cacheFile));
+            }
 
-            _logger.LogInformation("Downloading xmltv listings from {Path}", info.Path);
+            if (info.Path.StartsWith("http", StringComparison.OrdinalIgnoreCase))
+            {
+                _logger.LogInformation("Downloading xmltv listings from {Path}", info.Path);
 
-            Directory.CreateDirectory(Path.GetDirectoryName(cacheFile));
+                using var response = await _httpClientFactory.CreateClient(NamedClient.Default).GetAsync(info.Path, cancellationToken).ConfigureAwait(false);
+                await using var stream = await response.Content.ReadAsStreamAsync(cancellationToken).ConfigureAwait(false);
 
-            using var response = await _httpClientFactory.CreateClient(NamedClient.Default).GetAsync(info.Path, cancellationToken).ConfigureAwait(false);
-            await using var stream = await response.Content.ReadAsStreamAsync(cancellationToken).ConfigureAwait(false);
-            await using (var fileStream = new FileStream(cacheFile, FileMode.CreateNew, FileAccess.Write, FileShare.None, IODefaults.CopyToBufferSize, FileOptions.Asynchronous))
-            {
-                await stream.CopyToAsync(fileStream, cancellationToken).ConfigureAwait(false);
+                return await UnzipIfNeededAndCopy(info.Path, stream, cacheFile, cancellationToken).ConfigureAwait(false);
             }
+            else
+            {
+                await using var stream = new FileStream(info.Path, FileMode.Open, FileAccess.Read, FileShare.Read, IODefaults.FileStreamBufferSize, FileOptions.Asynchronous);
 
-            return UnzipIfNeeded(info.Path, cacheFile);
+                return await UnzipIfNeededAndCopy(info.Path, stream, cacheFile, cancellationToken).ConfigureAwait(false);
+            }
         }
 
-        private string UnzipIfNeeded(ReadOnlySpan<char> originalUrl, string file)
+        private async Task<string> UnzipIfNeededAndCopy(string originalUrl, Stream stream, string file, CancellationToken cancellationToken)
         {
-            ReadOnlySpan<char> ext = Path.GetExtension(originalUrl.LeftPart('?'));
+            string ext = Path.GetExtension(originalUrl.Split('?')[0]);
 
             if (ext.Equals(".gz", StringComparison.OrdinalIgnoreCase))
             {
                 try
                 {
-                    string tempFolder = ExtractGz(file);
-                    return FindXmlFile(tempFolder);
-                }
-                catch (Exception ex)
-                {
-                    _logger.LogError(ex, "Error extracting from gz file {File}", file);
-                }
-
-                try
-                {
-                    string tempFolder = ExtractFirstFileFromGz(file);
-                    return FindXmlFile(tempFolder);
+                    using var reader = new GZipStream(stream, CompressionMode.Decompress);
```

Comment:
I don't think this is comparable to the old code, this doesn't support gzipped tars AFAIK

Reply:
> I don't think this is comparable to the old code, this doesn't support gzipped tars AFAIK

I thought only xml and xml.gz files where supported, not xml.tar.gz. Is there any other types of files that should be supported? I see that files with .gz extension are decompressed, but not .tar ones

Reply:
The handling of tar files is located in the sharpcompress code. (The functions used in IZipClient)

Reply:
> The handling of tar files is located in the sharpcompress code. (The functions used in IZipClient)

I've been testing current code and the functions in ZipClient only decompress gzip (only SharpCompress's GZipReader is used and it doesn't handle tars). Are you sure this code is not comparable?

Reply:
Looks like you are right, it doesn't support tar.
Now I don't understand why ZipClient had 2 functions, `ExtractAllFromGz` and `ExtractFirstFileFromGz` as gzips can only contain 1 file AFAIK